crates/services/gas_price_service/src/v0/service.rs

```diff
@@ -0,0 +1,326 @@
+        &mut self,
+        height: u32,
+        gas_used: u64,
+        block_gas_capacity: u64,
+    ) -> anyhow::Result<()> {
+        let capacity = self.validate_block_gas_capacity(block_gas_capacity)?;
+
+        self.algorithm_updater
+            .update_l2_block_data(height, gas_used, capacity)?;
+
+        self.set_metadata().await?;
+        Ok(())
+    }
+
+    async fn apply_block_info_to_gas_algorithm(
+        &mut self,
+        l2_block: BlockInfo,
+    ) -> anyhow::Result<()> {
+        match l2_block {
+            BlockInfo::GenesisBlock => {
+                self.set_metadata().await?;
+            }
+            BlockInfo::Block {
+                height,
+                gas_used,
+                block_gas_capacity,
+            } => {
+                self.handle_normal_block(height, gas_used, block_gas_capacity)
+                    .await?;
+            }
+        }
+
+        self.update(self.algorithm_updater.algorithm()).await;
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl<L2, Metadata> RunnableService for GasPriceServiceV0<L2, Metadata>
+where
+    L2: L2BlockSource,
+    Metadata: MetadataStorage,
+{
+    const NAME: &'static str = "GasPriceServiceV0";
+    type SharedData = SharedV0Algorithm;
+    type Task = Self;
+    type TaskParams = ();
+
+    fn shared_data(&self) -> Self::SharedData {
+        self.shared_algo.clone()
+    }
+
+    async fn into_task(
+        mut self,
+        _state_watcher: &StateWatcher,
+        _params: Self::TaskParams,
+    ) -> anyhow::Result<Self::Task> {
+        let algorithm = self.algorithm_updater.algorithm();
+        self.shared_algo.update(algorithm).await;
+        Ok(self)
+    }
+}
+
+#[async_trait]
+impl<L2, Metadata> RunnableTask for GasPriceServiceV0<L2, Metadata>
+where
+    L2: L2BlockSource,
+    Metadata: MetadataStorage,
+{
+    async fn run(&mut self, watcher: &mut StateWatcher) -> anyhow::Result<bool> {
+        let should_continue;
+        tokio::select! {
+            biased;
+            _ = watcher.while_started() => {
+                tracing::debug!("Stopping gas price service");
+                should_continue = false;
+            }
+            l2_block_res = self.l2_block_source.get_l2_block() => {
+                tracing::info!("Received L2 block result: {:?}", l2_block_res);
+                let block = l2_block_res?;
+
+                tracing::debug!("Updating gas price algorithm");
+                self.apply_block_info_to_gas_algorithm(block).await?;
+                should_continue = true;
+            }
+        }
+        Ok(should_continue)
+    }
+
+    async fn shutdown(mut self) -> anyhow::Result<()> {
+        while let Some(Ok(block)) = self.l2_block_source.get_l2_block().now_or_never() {
+            tracing::debug!("Updating gas price algorithm");
+            self.apply_block_info_to_gas_algorithm(block).await?;
+        }
+        Ok(())
+    }
+}
+
+#[allow(clippy::arithmetic_side_effects)]
```

Comment:
for reviewers: these tests are also more or less the same, only thing changed is the signatures.